src/common/containers/GameAreaContainer.tsx

```diff
@@ -233,20 +238,43 @@ export class GameAreaContainer extends React.Component<GameAreaContainerProps, G
 
   public render(): JSX.Element {
     return (
-      <GameArea
-        {...this.props}
-        message={this.state.message}
-        onClickGameArea={this.handleClickGameArea}
-        onClickEndGame={this.handleClickEndGame}
-        onNextTutorialStep={this.handleNextTutorialStep}
-        onPrevTutorialStep={this.handlePrevTutorialStep}
-        onSelectTile={this.onSelectTile}
-      />
+      <>
+        {!inTest && <Prompt
+          when={this.shouldWarnBeforeLeavingPage}
+          message="Are you sure you want to leave the game?"
+        />}
+        <GameArea
+          {...this.props}
+          message={this.state.message}
+          onClickGameArea={this.handleClickGameArea}
+          onClickEndGame={this.handleClickEndGame}
+          onNextTutorialStep={this.handleNextTutorialStep}
+          onPrevTutorialStep={this.handlePrevTutorialStep}
+          onSelectTile={this.onSelectTile}
+        />
+      </>
     );
   }
 
+  private get shouldWarnBeforeLeavingPage(): boolean {
+    const { gameOver, isSpectator } = this.props;
+    return !gameOver && !isSpectator;
+  }
+
   private urlMatchesGameMode = (mode: string) => this.props.location.pathname.startsWith(urlForGameMode(mode));
 
+  private handleWindowBeforeUnload = (event: Event) => {
+    if (this.shouldWarnBeforeLeavingPage) {
+      // https://stackoverflow.com/a/25763325
+      event = event || window.event;
+      // Cancel the event as stated by the standard.
+      event.preventDefault();
+      // Chrome requires returnValue to be set.
+      (event as { returnValue: unknown }).returnValue = '';
```

Comment:
Huh I've always just done `event.preventDefault();`

Reply:
I guess "beforeunload" is different

Reply:
You will not believe how complicated beforeunload is, I had to do a bunch of research to get this to work